`Plug.CSRFProtection` refactoring

whatyouhide opened the pull request:
I refactored some code and tests for `Plug.CSRFProtection`. Most of the (more significant) changes have been documented in the diffs' comments. Let me know if everything is ok :smiley:.

josevalim commented on lib/plug/csrf_protection.ex:
```diff
@@ -51,18 +51,21 @@ defmodule Plug.CSRFProtection do
+
+    message =
+      "invalid CSRF (Cross Site Forgery Protection) token, make sure all "
+      <> "requests include a '_csrf_token' param or an 'x-csrf-token' header"
```
I think we should make it clear that only non-HEAD and non-GET requests as before. :)

josevalim commented:
I don't like most of the changes because they are syntax based. There isn't much refactoring but a lot of "let's use this syntax rather than this one". :(

josevalim commented on lib/plug/csrf_protection.ex:
```diff
@@ -106,26 +111,27 @@ defmodule Plug.CSRFProtection do
 
   defp verified_request?(conn, csrf_token) do
     conn.method in @unprotected_methods
+      || skip_csrf_protection?(conn)
```
The previous order was on purpose. We are checking by order that is most likely to match in production. The flag will only match in tests, so we moved to the bottom.

josevalim commented on lib/plug/csrf_protection.ex:
```diff
@@ -144,19 +155,18 @@ defmodule Plug.CSRFProtection do
-      conn
-    else
-      put_session(conn, "_csrf_token", current)
+    case Process.delete(:plug_csrf_token) do
```
:+1:

josevalim commented on lib/plug/csrf_protection.ex:
```diff
@@ -133,10 +139,15 @@ defmodule Plug.CSRFProtection do
-    conn.method == "GET" and not xhr?(conn) and not plug_skip_csrf_protection?(conn) and
-      Enum.any?(get_resp_header(conn, "content-type"),
-                &String.starts_with?(&1, ["text/javascript", "application/javascript"]))
+  defp cross_origin_js?(%Plug.Conn{method: "GET"} = conn),
```
:+1: for those changes but there is no need to add the parens around not (when we did not have it before).

josevalim commented:
If you are ok with it, let's keep the test changes and the ones in `cross_origin_js?` or `ensure_csrf_token`. The other ones are pretty much different syntax and there is no need for code churn. Thanks.

whatyouhide commented:
Hey @josevalim, thanks for the feedback. I actually tried to keep the syntax changes at a minimum; most of the changes are taking advantage of pattern matching to build multiple function heads instead of a function head with lost of conditionals/conditions in the body.

I'm reverting the order of the boolean expression in `verified_request?/2`, I didn't think of ordering them based on what happens more frequently in production. That's a great point!

As for the exception messages, I pretty much just tried to make them shorter and more concise. I decided to remove the mention to `GET` and `HEAD` requests mainly because it's incomplete (as the docs for `Plug.CSRFProtection`): as far as I understand, `OPTIONS` requests are unprotected too. I probably was wrong removing those mention and should have added a mention to `OPTIONS` instead. Let me know what you think, if you agree I'll update both the docs and revert the error message to mention `OPTIONS` requests.

whatyouhide commented:
Reverted the order in `verified_request?/2` and removed some stylistic changes (parens in `not` - sorry about that! - and newlines in `skip_csrf_protection?/1`.
